cpp/src/arrow/util/iterator.h

```diff
@@ -78,8 +78,35 @@ class VectorIterator : public Iterator<T> {
 };
 
 template <typename T>
-std::unique_ptr<VectorIterator<T>> MakeIterator(std::vector<T> v) {
+std::unique_ptr<Iterator<T>> MakeIterator(std::vector<T> v) {
   return std::unique_ptr<VectorIterator<T>>(new VectorIterator<T>(std::move(v)));
 }
 
+template <typename Fn, typename I, typename O = typename std::result_of<Fn(I)>::type>
+class MapIterator : public Iterator<O> {
+ public:
+  explicit MapIterator(Fn map, std::unique_ptr<Iterator<I>> it)
+      : map_(map), it_(std::move(it)) {}
```

Comment:
Why no `std::move` for `map`? Some callables, e.g. closures, may have a non-trivial payload.